src/mm/device_alloc.rs

```diff
@@ -27,9 +28,33 @@ unsafe impl Allocator for DeviceAlloc {
 		assert!(layout.align() <= BasePageSize::SIZE as usize);
 		let size = layout.size().align_up(BasePageSize::SIZE as usize);
 
-		let virt_addr = ptr.as_ptr().expose_provenance();
-		let phys_addr = PhysAddr::from(virt_addr);
+		let phys_addr = PhysAddr::from(self.phys_addr(ptr.as_ptr()));
 
 		super::physicalmem::deallocate(phys_addr, size);
 	}
 }
+
+impl DeviceAlloc {
+	/// Returns a pointer corresponding to `phys_addr`.
+	pub fn ptr<T>(&self, phys_addr: usize) -> *mut T {
+		ptr::with_exposed_provenance_mut(phys_addr + self.phys_offset())
+	}
+
+	/// Returns the physical address of `ptr`.
+	///
+	/// The address is only correct if `ptr` has been allocated by this allocator.
+	pub fn phys_addr<T: ?Sized>(&self, ptr: *mut T) -> usize {
+		ptr.expose_provenance() - self.phys_offset()
+	}
+
+	/// Returns the physical address offset.
+	///
+	/// This device allocator expects the complete physical memory to be mapped device-readable at this offset.
+	pub fn phys_offset(&self) -> usize {
```

Comment:
Make the function constant? I think virtualmem::kernel_heap_end() can also be made const, so it'd probably reduce the overhead a bit.